cron_jobs/caravan_cron.php

```diff
@@ -0,0 +1,131 @@
+            if ($system->environment === System::ENVIRONMENT_PROD) {
+                if (!isset($_POST['confirm']) || $_POST['confirm'] !== $production_key) {
+                    echo "WARNING PRODUCTION ENVIRONMENT! Confirm you would like to run this adhoc on <b>PRODUCTION ENVIRONMENT</b>!!!<br />
+                To confirm, type the following (case sensitive): $production_key<br />
+                <form action='{$system->router->base_url}/cron_jobs/caravan_cron.php?run_script=true' method='post'>
+                    <input type='text' name='confirm' /><input type='submit' value='Run Script' />
+                </form>";
+                } else {
+                    hourlyCaravan($system);
+                    $player->staff_manager->staffLog(
+                        StaffManager::STAFF_LOG_ADMIN,
+                        "{$player->user_name}({$player->user_id}) manually ran caravan cron."
+                    );
+                    echo "Script ran.";
+                }
+            } else {
+                hourlyCaravan($system);
+                $player->staff_manager->staffLog(
+                    StaffManager::STAFF_LOG_ADMIN,
+                    "{$player->user_name}({$player->user_id}) manually ran caravan cron."
+                );
+                echo "Script ran.";
+            }
+        } else {
+            echo "You can run the caravan cron script Adhoc. This is not reversible. <a href='{$system->router->base_url}/cron_jobs/caravan_cron.php?run_script=true'>Run</a>.";
+        }
+    }
+} else {
+    // Check for verify to run cron
+    $run_ok = false;
+    if (php_sapi_name() == 'cli') {
+        $run_ok = true;
+    } else if ($_SERVER['REMOTE_ADDR'] == $_SERVER['SERVER_ADDR']) {
+        $run_ok = true;
+    }
+
+    if ($run_ok) {
+        hourlyCaravan($system);
+        $system->log('cron', 'Hourly Caravan', "Caravans and resources have been processed.");
+    } else {
+        $system->log('cron', 'Invalid access', "Attempted access by " . $_SERVER['REMOTE_ADDR']);
+    }
+}
+
+function hourlyCaravan(System $system, $debug = true): void
+{
+    /* step 1: update village resources */
+
+    // get villages
+    $village_result = $system->db->query("SELECT * FROM `villages`");
+    $village_result = $system->db->fetch_all($village_result);
+    $villages = [];
+    foreach ($village_result as $village) {
+        $villages[$village['village_id']] = new Village($system, village_row: $village);
+    }
+
+    // get resources from caravans
+    $caravan_result = $system->db->query("SELECT * FROM `caravans`");
+    $caravan_result = $system->db->fetch_all($caravan_result);
+    foreach ($caravan_result as $caravan) {
+        $caravan_resources = json_decode($caravan['resources'], true);
+        // add resource to village
+        foreach ($caravan_resources as $resource_id => $quantity) {
+            $villages[$caravan['village_id']]->addResource($resource_id, $quantity);
+        }
```

Comment:
That's correct, since it runs every hour it assumes all caravans have now reached their destination.
If we need to rework this at some point, we can use the stop time of the caravan (they have a set start and duration so determining end time is how we know when to stop displaying them).